<!doctype html><html lang=en class=no-js> <head><meta charset=utf-8><meta name=viewport content="width=device-width,initial-scale=1"><meta name=description content="An OpenOmics Viral Metagenomics Pipeline"><meta name=author content="Skyler Kuhn"><link rel=icon href=../../assets/favicon/favicon.ico><meta name=generator content="mkdocs-1.1.2, mkdocs-material-7.2.3"><title>metavirs run - metavirs</title><link rel=stylesheet href=../../assets/stylesheets/main.f7f47774.min.css><link rel=stylesheet href=../../assets/stylesheets/palette.3f5d1f46.min.css><link rel=preconnect href=https://fonts.gstatic.com crossorigin><link rel=stylesheet href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback"><style>:root{--md-text-font-family:"Roboto";--md-code-font-family:"Roboto Mono"}</style><link rel=stylesheet href=../../css/extra.css></head> <body dir=ltr data-md-color-scheme=default data-md-color-primary data-md-color-accent> <script>function __prefix(e){return new URL("../..",location).pathname+"."+e}function __get(e,t=localStorage){return JSON.parse(t.getItem(__prefix(e)))}</script> <script>var palette=__get("__palette");if(null!==palette&&"object"==typeof palette.color)for(var key in palette.color)document.body.setAttribute("data-md-color-"+key,palette.color[key])</script> <input class=md-toggle data-md-toggle=drawer type=checkbox id=__drawer autocomplete=off> <input class=md-toggle data-md-toggle=search type=checkbox id=__search autocomplete=off> <label class=md-overlay for=__drawer></label> <div data-md-component=skip> <a href=#metavirs-run class=md-skip> Skip to content </a> </div> <div data-md-component=announce> </div> <header class=md-header data-md-component=header> <nav class="md-header__inner md-grid" aria-label=Header> <a href=../.. title=metavirs class="md-header__button md-logo" aria-label=metavirs data-md-component=logo> <img src=../../assets/icons/doc-book.svg alt=logo> </a> <label class="md-header__button md-icon" for=__drawer> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M3 6h18v2H3V6m0 5h18v2H3v-2m0 5h18v2H3v-2z"/></svg> </label> <div class=md-header__title data-md-component=header-title> <div class=md-header__ellipsis> <div class=md-header__topic> <span class=md-ellipsis> metavirs </span> </div> <div class=md-header__topic data-md-component=header-topic> <span class=md-ellipsis> metavirs run </span> </div> </div> </div> <form class=md-header__option data-md-component=palette> <input class=md-option data-md-color-media data-md-color-scheme=default data-md-color-primary data-md-color-accent aria-label="Switch to dark mode" type=radio name=__palette id=__palette_1> <label class="md-header__button md-icon" title="Switch to dark mode" for=__palette_2 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M12 6a6 6 0 0 1 6 6c0 2.22-1.21 4.16-3 5.2V19a1 1 0 0 1-1 1h-4a1 1 0 0 1-1-1v-1.8c-1.79-1.04-3-2.98-3-5.2a6 6 0 0 1 6-6m2 15v1a1 1 0 0 1-1 1h-2a1 1 0 0 1-1-1v-1h4m6-10h3v2h-3v-2M1 11h3v2H1v-2M13 1v3h-2V1h2M4.92 3.5l2.13 2.14-1.42 1.41L3.5 4.93 4.92 3.5m12.03 2.13 2.12-2.13 1.43 1.43-2.13 2.12-1.42-1.42z"/></svg> </label> <input class=md-option data-md-color-media data-md-color-scheme=slate data-md-color-primary data-md-color-accent aria-label="Switch to light mode" type=radio name=__palette id=__palette_2> <label class="md-header__button md-icon" title="Switch to light mode" for=__palette_1 hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="m17.75 4.09-2.53 1.94.91 3.06-2.63-1.81-2.63 1.81.91-3.06-2.53-1.94L12.44 4l1.06-3 1.06 3 3.19.09m3.5 6.91-1.64 1.25.59 1.98-1.7-1.17-1.7 1.17.59-1.98L15.75 11l2.06-.05L18.5 9l.69 1.95 2.06.05m-2.28 4.95c.83-.08 1.72 1.1 1.19 1.85-.32.45-.66.87-1.08 1.27C15.17 23 8.84 23 4.94 19.07c-3.91-3.9-3.91-10.24 0-14.14.4-.4.82-.76 1.27-1.08.75-.53 1.93.36 1.85 1.19-.27 2.86.69 5.83 2.89 8.02a9.96 9.96 0 0 0 8.02 2.89m-1.64 2.02a12.08 12.08 0 0 1-7.8-3.47c-2.17-2.19-3.33-5-3.49-7.82-2.81 3.14-2.7 7.96.31 10.98 3.02 3.01 7.84 3.12 10.98.31z"/></svg> </label> </form> <label class="md-header__button md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg> </label> <div class=md-search data-md-component=search role=dialog> <label class=md-search__overlay for=__search></label> <div class=md-search__inner role=search> <form class=md-search__form name=search> <input type=text class=md-search__input name=query aria-label=Search placeholder=Search autocapitalize=off autocorrect=off autocomplete=off spellcheck=false data-md-component=search-query required> <label class="md-search__icon md-icon" for=__search> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.516 6.516 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5z"/></svg> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg> </label> <nav class=md-search__options aria-label=Search> <button type=reset class="md-search__icon md-icon" aria-label=Clear tabindex=-1> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12 19 6.41z"/></svg> </button> </nav> </form> <div class=md-search__output> <div class=md-search__scrollwrap data-md-scrollfix> <div class=md-search-result data-md-component=search-result> <div class=md-search-result__meta> Initializing search </div> <ol class=md-search-result__list></ol> </div> </div> </div> </div> </div> <div class=md-header__source> <a href=https://github.com/OpenOmics/metavirs/ title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> OpenOmics/metavirs </div> </a> </div> </nav> </header> <div class=md-container data-md-component=container> <nav class=md-tabs aria-label=Tabs data-md-component=tabs> <div class="md-tabs__inner md-grid"> <ul class=md-tabs__list> <li class=md-tabs__item> <a href=../.. class=md-tabs__link> About </a> </li> <li class=md-tabs__item> <a href=./ class="md-tabs__link md-tabs__link--active"> Commands </a> </li> <li class=md-tabs__item> <a href=../../faq/questions/ class=md-tabs__link> FAQ </a> </li> <li class=md-tabs__item> <a href=../../license/ class=md-tabs__link> License </a> </li> </ul> </div> </nav> <main class=md-main data-md-component=main> <div class="md-main__inner md-grid"> <div class="md-sidebar md-sidebar--primary" data-md-component=sidebar data-md-type=navigation> <div class=md-sidebar__scrollwrap> <div class=md-sidebar__inner> <nav class="md-nav md-nav--primary md-nav--lifted md-nav--integrated" aria-label=Navigation data-md-level=0> <label class=md-nav__title for=__drawer> <a href=../.. title=metavirs class="md-nav__button md-logo" aria-label=metavirs data-md-component=logo> <img src=../../assets/icons/doc-book.svg alt=logo> </a> metavirs </label> <div class=md-nav__source> <a href=https://github.com/OpenOmics/metavirs/ title="Go to repository" class=md-source data-md-component=source> <div class="md-source__icon md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 448 512"><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81z"/></svg> </div> <div class=md-source__repository> OpenOmics/metavirs </div> </a> </div> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../.. class=md-nav__link> About </a> </li> <li class="md-nav__item md-nav__item--active md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_2 type=checkbox id=__nav_2 checked> <label class=md-nav__link for=__nav_2> Commands <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=Commands data-md-level=1> <label class=md-nav__title for=__nav_2> <span class="md-nav__icon md-icon"></span> Commands </label> <ul class=md-nav__list data-md-scrollfix> <li class="md-nav__item md-nav__item--active"> <input class="md-nav__toggle md-toggle" data-md-toggle=toc type=checkbox id=__toc> <label class="md-nav__link md-nav__link--active" for=__toc> metavirs run <span class="md-nav__icon md-icon"></span> </label> <a href=./ class="md-nav__link md-nav__link--active"> metavirs run </a> <nav class="md-nav md-nav--secondary" aria-label="Table of contents"> <label class=md-nav__title for=__toc> <span class="md-nav__icon md-icon"></span> Table of contents </label> <ul class=md-nav__list data-md-component=toc data-md-scrollfix> <li class=md-nav__item> <a href=#1-about class=md-nav__link> 1. About </a> </li> <li class=md-nav__item> <a href=#2-synopsis class=md-nav__link> 2. Synopsis </a> <nav class=md-nav aria-label="2. Synopsis"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#21-required-arguments class=md-nav__link> 2.1 Required Arguments </a> </li> <li class=md-nav__item> <a href=#22-analysis-options class=md-nav__link> 2.2 Analysis options </a> </li> <li class=md-nav__item> <a href=#23-orchestration-options class=md-nav__link> 2.3 Orchestration options </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=#24-miscellaneous-options class=md-nav__link> 2.4 Miscellaneous options </a> </li> <li class=md-nav__item> <a href=#3-examples class=md-nav__link> 3. Examples </a> <nav class=md-nav aria-label="3. Examples"> <ul class=md-nav__list> <li class=md-nav__item> <a href=#31-biowulf class=md-nav__link> 3.1 Biowulf </a> </li> <li class=md-nav__item> <a href=#32-bigsky class=md-nav__link> 3.2 BigSky </a> </li> </ul> </nav> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../unlock/ class=md-nav__link> metavirs unlock </a> </li> <li class=md-nav__item> <a href=../install/ class=md-nav__link> metavirs install </a> </li> <li class=md-nav__item> <a href=../cache/ class=md-nav__link> metavirs cache </a> </li> </ul> </nav> </li> <li class="md-nav__item md-nav__item--nested"> <input class="md-nav__toggle md-toggle" data-md-toggle=__nav_3 type=checkbox id=__nav_3> <label class=md-nav__link for=__nav_3> FAQ <span class="md-nav__icon md-icon"></span> </label> <nav class=md-nav aria-label=FAQ data-md-level=1> <label class=md-nav__title for=__nav_3> <span class="md-nav__icon md-icon"></span> FAQ </label> <ul class=md-nav__list data-md-scrollfix> <li class=md-nav__item> <a href=../../faq/questions/ class=md-nav__link> General Questions </a> </li> </ul> </nav> </li> <li class=md-nav__item> <a href=../../license/ class=md-nav__link> License </a> </li> </ul> </nav> </div> </div> </div> <div class=md-content data-md-component=content> <article class="md-content__inner md-typeset"> <a href=https://github.com/OpenOmics/metavirs/edit/main/docs/usage/run.md title="Edit this page" class="md-content__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20.71 7.04c.39-.39.39-1.04 0-1.41l-2.34-2.34c-.37-.39-1.02-.39-1.41 0l-1.84 1.83 3.75 3.75M3 17.25V21h3.75L17.81 9.93l-3.75-3.75L3 17.25z"/></svg> </a> <h1 id=metavirs-run><code>metavirs <b>run</b></code><a class=headerlink href=#metavirs-run title="Permanent link">&para;</a></h1> <h2 id=1-about>1. About<a class=headerlink href=#1-about title="Permanent link">&para;</a></h2> <p>The <code>metavirs</code> executable is composed of several inter-related sub commands. Please see <code>metavirs -h</code> for all available options.</p> <p>This part of the documentation describes options and concepts for <code>metavirs <b>run</b></code> sub command in more detail. With minimal configuration, the <strong><code>run</code></strong> sub command enables you to start running metavirs pipeline. </p> <p>Setting up the metavirs pipeline is fast and easy! In its most basic form, <code>metavirs <b>run</b></code> only has <em>two required inputs</em>.</p> <h2 id=2-synopsis>2. Synopsis<a class=headerlink href=#2-synopsis title="Permanent link">&para;</a></h2> <div class=highlight><pre><span></span><code>$ metavirs run [--help] [--aggregate] [--length-filter LENGTH_FILTER] \
     [--mode &lt;slurm,local&gt;] [--job-name JOB_NAME] \
     [--dry-run] [--silent] [--sif-cache SIF_CACHE] \
     [--singularity-cache SINGULARITY_CACHE] \
     [--tmp-dir TMP_DIR] [--threads THREADS] \
     [--resource-bundle RESOURCE_BUNDLE] \
      --input INPUT [INPUT ...] \
      --output OUTPUT
</code></pre></div> <p>The synopsis for each command shows its arguments and their usage. Optional arguments are shown in square brackets.</p> <p>A user <strong>must</strong> provide a list of FastQ (globbing is supported) to analyze via <code>--input</code> argument and an output directory to store results via <code>--output</code> argument.</p> <p>Use you can always use the <code>-h</code> option for information on a specific command. </p> <h3 id=21-required-arguments>2.1 Required Arguments<a class=headerlink href=#21-required-arguments title="Permanent link">&para;</a></h3> <p>Each of the following arguments are required. Failure to provide a required argument will result in a non-zero exit-code.</p> <p><code>--input INPUT [INPUT ...]</code> </p> <blockquote> <p><strong>Input FastQ file(s).</strong><br> <em>type: file(s)</em> </p> <p>One or more FastQ files can be provided. The pipeline supports single-end and paired-end data. Single-end and paired-end FastQ files can be provided to this option simultaneously. Under the hood, the pipeline will run the correct set of rules and options for a given sample. When providing each input file, please seperate each file with a space. Globbing is supported! This makes selecting FastQ files easy. Note: Input FastQ files must be gzipp-ed.</p> <p><strong><em>Example:</em></strong> <code>--input .tests/*.R?.fastq.gz</code></p> </blockquote> <hr> <p><code>--output OUTPUT</code></p> <blockquote> <p><strong>Path to an output directory.</strong> <br> <em>type: path</em></p> <p>This location is where the pipeline will create all of its output files, also known as the pipeline's working directory. If the provided output directory does not exist, it will be created automatically.</p> <p><strong><em>Example:</em></strong> <code>--output /data/$USER/metavirs_out</code></p> </blockquote> <h3 id=22-analysis-options>2.2 Analysis options<a class=headerlink href=#22-analysis-options title="Permanent link">&para;</a></h3> <p>Each of the following arguments are optional, and do not need to be provided. </p> <p><code>--aggregate</code> </p> <blockquote> <p><strong>Aggregate results into one report.</strong><br> <em>type: boolean flag</em></p> <p>Aggregates contig annotation results into one mutli-sample, project-level interactive Krona report. By default, any resulting reports will be created at a per-sample level.</p> <p><strong><em>Example:</em></strong> <code>--aggregate</code></p> </blockquote> <hr> <p><code>--length-filter LENGTH_FILTER</code> </p> <blockquote> <p><strong>Filter contigs by total length (bp).</strong><br> <em>type: integer</em> <em>default: 500</em></p> <p>The contig length filter is used to remove any annotated contigs less than this threshold. Annotating small contigs can lead to lower-confidence classifications. This filter is only applied to metaspades contigs.</p> <p><strong><em>Example:</em></strong> <code>--length-filter 750</code></p> </blockquote> <h3 id=23-orchestration-options>2.3 Orchestration options<a class=headerlink href=#23-orchestration-options title="Permanent link">&para;</a></h3> <p>Each of the following arguments are optional, and do not need to be provided. </p> <p><code>--dry-run</code> </p> <blockquote> <p><strong>Dry run the pipeline.</strong><br> <em>type: boolean flag</em></p> <p>Displays what steps in the pipeline remain or will be run. Does not execute anything!</p> <p><strong><em>Example:</em></strong> <code>--dry-run</code></p> </blockquote> <hr> <p><code>--silent</code> </p> <blockquote> <p><strong>Silence standard output.</strong><br> <em>type: boolean flag</em></p> <p>Reduces the amount of information directed to standard output when submitting master job to the job scheduler. Only the job id of the master job is returned.</p> <p><strong><em>Example:</em></strong> <code>--silent</code></p> </blockquote> <hr> <p><code>--mode {slurm,local}</code> </p> <blockquote> <p><strong>Execution Method.</strong><br> <em>type: string</em><br> <em>default: slurm</em></p> <p>Execution Method. Defines the mode or method of execution. Vaild mode options include: slurm or local. </p> <p><strong><em>slurm</em></strong> <br> The slurm execution method will submit jobs to the <a href=https://slurm.schedmd.com/ >SLURM workload manager</a>. It is recommended running metavirs in this mode as execution will be significantly faster in a distributed environment. This is the default mode of execution.</p> <p><strong><em>local</em></strong><br> Local executions will run serially on compute instance. This is useful for testing, debugging, or when a users does not have access to a high performance computing environment. If this option is not provided, it will default to a local execution mode. </p> <p><strong><em>Example:</em></strong> <code>--mode slurm</code></p> </blockquote> <hr> <p><code>--job-name JOB_NAME</code> </p> <blockquote> <p><strong>Set the name of the pipeline's master job.</strong><br> <em>type: string</em> <em>default: pl:metavirs</em></p> <p>When submitting the pipeline to a job scheduler, like SLURM, this option always you to set the name of the pipeline's master job. By default, the name of the pipeline's master job is set to "pl:metavirs".</p> <p><strong><em>Example:</em></strong> <code>--job-name pl_id-42</code></p> </blockquote> <hr> <p><code>--singularity-cache SINGULARITY_CACHE</code> </p> <blockquote> <p><strong>Overrides the $SINGULARITY_CACHEDIR environment variable.</strong><br> <em>type: path</em><br> <em>default: <code>--output OUTPUT/.singularity</code></em></p> <p>Singularity will cache image layers pulled from remote registries. This ultimately speeds up the process of pull an image from DockerHub if an image layer already exists in the singularity cache directory. By default, the cache is set to the value provided to the <code>--output</code> argument. Please note that this cache cannot be shared across users. Singularity strictly enforces you own the cache directory and will return a non-zero exit code if you do not own the cache directory! See the <code>--sif-cache</code> option to create a shareable resource. </p> <p><strong><em>Example:</em></strong> <code>--singularity-cache /data/$USER/.singularity</code></p> </blockquote> <hr> <p><code>--sif-cache SIF_CACHE</code></p> <blockquote> <p><strong>Path where a local cache of SIFs are stored.</strong><br> <em>type: path</em> </p> <p>Uses a local cache of SIFs on the filesystem. This SIF cache can be shared across users if permissions are set correctly. If a SIF does not exist in the SIF cache, the image will be pulled from Dockerhub and a warning message will be displayed. The <code>metavirs cache</code> subcommand can be used to create a local SIF cache. Please see <code>metavirs cache</code> for more information. This command is extremely useful for avoiding DockerHub pull rate limits. It also remove any potential errors that could occur due to network issues or DockerHub being temporarily unavailable. We recommend running metavirs with this option when ever possible.</p> <p><strong><em>Example:</em></strong> <code>--singularity-cache /data/$USER/SIFs</code></p> </blockquote> <hr> <p><code>--resource-bundle RESOURCE_BUNDLE</code></p> <blockquote> <p><strong>Path to a resource bundle downloaded with the install sub command.</strong><br> <em>type: path</em> </p> <p>The resource bundle contains the set of required reference files for processing any data. The path provided to this option will be the path to the metavirs directory that was created when running the install sub command. Please see the install sub command for more information about downloading the pipeline's resource bundle.</p> <p><strong><em>Example:</em></strong> <code>--resource-bundle /data/$USER/refs/metavirs</code></p> </blockquote> <hr> <p><code>--threads THREADS</code> </p> <blockquote> <p><strong>Max number of threads for each process.</strong><br> <em>type: integer</em><br> <em>default: 2</em></p> <p>Max number of threads for each process. This option is more applicable when running the pipeline with <code>--mode local</code>. It is recommended setting this vaule to the maximum number of CPUs available on the host machine.</p> <p><strong><em>Example:</em></strong> <code>--threads 12</code></p> </blockquote> <hr> <p><code>--tmp-dir TMP_DIR</code> </p> <blockquote> <p><strong>Max number of threads for each process.</strong><br> <em>type: path</em><br> <em>default: <code>/tmp/</code></em></p> <p>Path on the file system for writing temporary output files. By default, the temporary directory is set to '/tmp/' for increased compatibility; however, if you are running the pipeline on a target system with a dedicated scratch space, this option will need to be specified. Ideally, this path should point to a dedicated location on the filesystem for writing tmp files. On many systems, this location is set to somewhere in /scratch. If you need to inject a variable into this string that should NOT be expanded, please quote this options value in single quotes.</p> <p><strong><em>Example:</em></strong> <code>--tmp-dir /scratch/$USER/</code></p> </blockquote> <h2 id=24-miscellaneous-options>2.4 Miscellaneous options<a class=headerlink href=#24-miscellaneous-options title="Permanent link">&para;</a></h2> <p>Each of the following arguments are optional, and do not need to be provided. </p> <p><code>-h, --help</code> </p> <blockquote> <p><strong>Display Help.</strong><br> <em>type: boolean flag</em></p> <p>Shows command's synopsis, help message, and an example command</p> <p><strong><em>Example:</em></strong> <code>--help</code></p> </blockquote> <h2 id=3-examples>3. Examples<a class=headerlink href=#3-examples title="Permanent link">&para;</a></h2> <h3 id=31-biowulf>3.1 Biowulf<a class=headerlink href=#31-biowulf title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><span class=c1># Step 0.) Grab an interactive node,</span>
<span class=c1># do not run on head node if you are</span>
<span class=c1># on a cluster, like Biowulf/BigSky!</span>
srun -N <span class=m>1</span> -n <span class=m>1</span> --time<span class=o>=</span><span class=m>1</span>:00:00 --mem<span class=o>=</span>8gb --cpus-per-task<span class=o>=</span><span class=m>2</span> --pty bash
<span class=c1># Add any missing dependencies to $PATH</span>
module purge
module load singularity
module load snakemake

<span class=c1># Step 1A.) Dry-run the pipeline,</span>
<span class=c1># this will display what steps will </span>
<span class=c1># be run or what steps still remain.</span>
./metavirs run --input .tests/*.gz <span class=se>\</span>
    --output /data/<span class=nv>$USER</span>/metavirs_out <span class=se>\</span>
    --mode slurm <span class=se>\</span>
    --aggregate <span class=se>\</span>
    --length-filter <span class=m>500</span> <span class=se>\</span>
    --dry-run

<span class=c1># Step 1B.) Run the viral metagenomics</span>
<span class=c1># pipeline. The slurm mode will submit </span>
<span class=c1># jobs to the cluster. We recommended </span>
<span class=c1># running metavirs in this mode, if</span>
<span class=c1># you have access to a cluster like</span>
<span class=c1># Biowulf or BigSky.</span>
./metavirs run --input .tests/*.gz <span class=se>\</span>
    --output /data/<span class=nv>$USER</span>/metavirs_out <span class=se>\</span>
    --mode slurm <span class=se>\</span>
    --aggregate <span class=se>\</span>
    --length-filter <span class=m>500</span>
</code></pre></div> <h3 id=32-bigsky>3.2 BigSky<a class=headerlink href=#32-bigsky title="Permanent link">&para;</a></h3> <div class=highlight><pre><span></span><code><span class=c1># Step 0.) Grab an interactive node,</span>
<span class=c1># do not run on head node if you are</span>
<span class=c1># on a cluster, like Biowulf/BigSky!</span>
srun -N <span class=m>1</span> -n <span class=m>1</span> --time<span class=o>=</span><span class=m>1</span>:00:00 --mem<span class=o>=</span>8gb --cpus-per-task<span class=o>=</span><span class=m>2</span> --pty bash

<span class=c1># Add any missing dependencies to $PATH,</span>
<span class=c1># adds singularity, snakemake, and metavirs</span>
module purge
<span class=nb>source</span> /gs1/RTS/OpenOmics/bin/dependencies.sh

<span class=c1># Step 1A.) Dry-run the pipeline,</span>
<span class=c1># this will display what steps will </span>
<span class=c1># be run or what steps still remain.</span>
metavirs run --input .tests/*.gz <span class=se>\</span>
    --output /gs1/Scratch/<span class=nv>$USER</span>/metavirs_out <span class=se>\</span>
    --mode slurm <span class=se>\</span>
    --sif-cache /gs1/RTS/OpenOmics/SIFs/ <span class=se>\</span>
    --resource-bundle /gs1/RTS/OpenOmics/references/metavirs/ <span class=se>\</span>
    --aggregate <span class=se>\</span>
    --length-filter <span class=m>500</span> <span class=se>\</span>
    --dry-run

<span class=c1># Step 1B.) Run the viral metagenomics</span>
<span class=c1># pipeline. The slurm mode will submit </span>
<span class=c1># jobs to the cluster. We recommended </span>
<span class=c1># running metavirs in this mode, if</span>
<span class=c1># you have access to a cluster like</span>
<span class=c1># Biowulf or BigSky.</span>
metavirs run --input .tests/*.gz <span class=se>\</span>
    --output /gs1/Scratch/<span class=nv>$USER</span>/metavirs_out <span class=se>\</span>
    --mode slurm <span class=se>\</span>
    --sif-cache /gs1/RTS/OpenOmics/SIFs/ <span class=se>\</span>
    --resource-bundle /gs1/RTS/OpenOmics/references/metavirs/ <span class=se>\</span>
    --aggregate <span class=se>\</span>
    --length-filter <span class=m>500</span>
</code></pre></div> <hr> <div class=md-source-date> <small> Last update: 2025-03-17 </small> </div> </article> </div> </div> <a href=# class="md-top md-icon" data-md-component=top data-md-state=hidden> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"/></svg> Back to top </a> </main> <footer class=md-footer> <nav class="md-footer__inner md-grid" aria-label=Footer> <a href=../.. class="md-footer__link md-footer__link--prev" aria-label="Previous: About" rel=prev> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11h12z"/></svg> </div> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Previous </span> About </div> </div> </a> <a href=../unlock/ class="md-footer__link md-footer__link--next" aria-label="Next: metavirs unlock" rel=next> <div class=md-footer__title> <div class=md-ellipsis> <span class=md-footer__direction> Next </span> metavirs unlock </div> </div> <div class="md-footer__button md-icon"> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 24 24"><path d="M4 11v2h12l-5.5 5.5 1.42 1.42L19.84 12l-7.92-7.92L10.5 5.5 16 11H4z"/></svg> </div> </a> </nav> <div class="md-footer-meta md-typeset"> <div class="md-footer-meta__inner md-grid"> <div class=md-footer-copyright> <div class=md-footer-copyright__highlight> Copyright &copy; 2023 OpenOmics </div> Made with <a href=https://squidfunk.github.io/mkdocs-material/ target=_blank rel=noopener> Material for MkDocs </a> </div> <div class=md-footer-social> <a href=https://idss-bioinformatics.nih.gov/ target=_blank rel=noopener title=idss-bioinformatics.nih.gov class=md-footer-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><path d="M96 224c35.3 0 64-28.7 64-64s-28.7-64-64-64-64 28.7-64 64 28.7 64 64 64zm448 0c35.3 0 64-28.7 64-64s-28.7-64-64-64-64 28.7-64 64 28.7 64 64 64zm32 32h-64c-17.6 0-33.5 7.1-45.1 18.6 40.3 22.1 68.9 62 75.1 109.4h66c17.7 0 32-14.3 32-32v-32c0-35.3-28.7-64-64-64zm-256 0c61.9 0 112-50.1 112-112S381.9 32 320 32 208 82.1 208 144s50.1 112 112 112zm76.8 32h-8.3c-20.8 10-43.9 16-68.5 16s-47.6-6-68.5-16h-8.3C179.6 288 128 339.6 128 403.2V432c0 26.5 21.5 48 48 48h288c26.5 0 48-21.5 48-48v-28.8c0-63.6-51.6-115.2-115.2-115.2zm-223.7-13.4C161.5 263.1 145.6 256 128 256H64c-35.3 0-64 28.7-64 64v32c0 17.7 14.3 32 32 32h65.9c6.3-47.4 34.9-87.3 75.2-109.4z"/></svg> </a> <a href=https://github.com/OpenOmics target=_blank rel=noopener title=github.com class=md-footer-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 496 512"><path d="M165.9 397.4c0 2-2.3 3.6-5.2 3.6-3.3.3-5.6-1.3-5.6-3.6 0-2 2.3-3.6 5.2-3.6 3-.3 5.6 1.3 5.6 3.6zm-31.1-4.5c-.7 2 1.3 4.3 4.3 4.9 2.6 1 5.6 0 6.2-2s-1.3-4.3-4.3-5.2c-2.6-.7-5.5.3-6.2 2.3zm44.2-1.7c-2.9.7-4.9 2.6-4.6 4.9.3 2 2.9 3.3 5.9 2.6 2.9-.7 4.9-2.6 4.6-4.6-.3-1.9-3-3.2-5.9-2.9zM244.8 8C106.1 8 0 113.3 0 252c0 110.9 69.8 205.8 169.5 239.2 12.8 2.3 17.3-5.6 17.3-12.1 0-6.2-.3-40.4-.3-61.4 0 0-70 15-84.7-29.8 0 0-11.4-29.1-27.8-36.6 0 0-22.9-15.7 1.6-15.4 0 0 24.9 2 38.6 25.8 21.9 38.6 58.6 27.5 72.9 20.9 2.3-16 8.8-27.1 16-33.7-55.9-6.2-112.3-14.3-112.3-110.5 0-27.5 7.6-41.3 23.6-58.9-2.6-6.5-11.1-33.3 2.6-67.9 20.9-6.5 69 27 69 27 20-5.6 41.5-8.5 62.8-8.5s42.8 2.9 62.8 8.5c0 0 48.1-33.6 69-27 13.7 34.7 5.2 61.4 2.6 67.9 16 17.7 25.8 31.5 25.8 58.9 0 96.5-58.9 104.2-114.8 110.5 9.2 7.9 17 22.9 17 46.4 0 33.7-.3 75.4-.3 83.6 0 6.5 4.6 14.4 17.3 12.1C428.2 457.8 496 362.9 496 252 496 113.3 383.5 8 244.8 8zM97.2 352.9c-1.3 1-1 3.3.7 5.2 1.6 1.6 3.9 2.3 5.2 1 1.3-1 1-3.3-.7-5.2-1.6-1.6-3.9-2.3-5.2-1zm-10.8-8.1c-.7 1.3.3 2.9 2.3 3.9 1.6 1 3.6.7 4.3-.7.7-1.3-.3-2.9-2.3-3.9-2-.6-3.6-.3-4.3.7zm32.4 35.6c-1.6 1.3-1 4.3 1.3 6.2 2.3 2.3 5.2 2.6 6.5 1 1.3-1.3.7-4.3-1.3-6.2-2.2-2.3-5.2-2.6-6.5-1zm-11.4-14.7c-1.6 1-1.6 3.6 0 5.9 1.6 2.3 4.3 3.3 5.6 2.3 1.6-1.3 1.6-3.9 0-6.2-1.4-2.3-4-3.3-5.6-2z"/></svg> </a> <a href=https://hub.docker.com/u/skchronicles target=_blank rel=noopener title=hub.docker.com class=md-footer-social__link> <svg xmlns=http://www.w3.org/2000/svg viewbox="0 0 640 512"><path d="M349.9 236.3h-66.1v-59.4h66.1v59.4zm0-204.3h-66.1v60.7h66.1V32zm78.2 144.8H362v59.4h66.1v-59.4zm-156.3-72.1h-66.1v60.1h66.1v-60.1zm78.1 0h-66.1v60.1h66.1v-60.1zm276.8 100c-14.4-9.7-47.6-13.2-73.1-8.4-3.3-24-16.7-44.9-41.1-63.7l-14-9.3-9.3 14c-18.4 27.8-23.4 73.6-3.7 103.8-8.7 4.7-25.8 11.1-48.4 10.7H2.4c-8.7 50.8 5.8 116.8 44 162.1 37.1 43.9 92.7 66.2 165.4 66.2 157.4 0 273.9-72.5 328.4-204.2 21.4.4 67.6.1 91.3-45.2 1.5-2.5 6.6-13.2 8.5-17.1l-13.3-8.9zm-511.1-27.9h-66v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1v-59.4zm78.1 0h-66.1v59.4h66.1v-59.4zm-78.1-72.1h-66.1v60.1h66.1v-60.1z"/></svg> </a> </div> </div> </div> </footer> </div> <div class=md-dialog data-md-component=dialog> <div class="md-dialog__inner md-typeset"></div> </div> <script id=__config type=application/json>{"base": "../..", "features": ["navigation.tabs", "navigation.top", "toc.integrate"], "translations": {"clipboard.copy": "Copy to clipboard", "clipboard.copied": "Copied to clipboard", "search.config.lang": "en", "search.config.pipeline": "trimmer, stopWordFilter", "search.config.separator": "[\\s\\-]+", "search.placeholder": "Search", "search.result.placeholder": "Type to start searching", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.term.missing": "Missing", "select.version.title": "Select version"}, "search": "../../assets/javascripts/workers/search.709b4209.min.js", "version": {"provider": "mike"}}</script> <script src=../../assets/javascripts/bundle.29db7785.min.js></script> </body> </html>